### ISSUE TITLE
gdc on ARM, syntax errors building enumap.d

### ISSUE THREAD

**Issue by SimonN** [rcorre/enumap#1]
A friend tried to build my D/dub project on an ARM machine using gdc, because dmd doesn't seem to be available for that architecture. Here are some syntax errors in enumap by gdc.

I'm sorry that I can't test anything with gdc now -- the ARM machine isn't mine, and my dev machine would only run gdc 4.3. >_> I guess that moving the user-defined attributes is enough to fix 3 of the 4 errors. If I get hold of the friend soon enough, I'll try to make a pull request for you. :-)

gdc (Ubuntu/Linaro 4.9.2-18ubuntu13) 4.9.2

```
Performing "debug" build using gdc for arm, arm_thumb, arm_hardfloat.
enumap 0.4.0: building configuration "library"...
../.dub/packages/enumap-0.4.0/src/enumap.d:220: error: user defined attributes cannot appear as postfixes
../.dub/packages/enumap-0.4.0/src/enumap.d:220: error: unexpected identifer 'dg' in declarator
../.dub/packages/enumap-0.4.0/src/enumap.d:220: error: found ';' when expecting ')'
../.dub/packages/enumap-0.4.0/src/enumap.d:222: error: found 'res' when expecting '.' following int
../.dub/packages/enumap-0.4.0/src/enumap.d:222: error: found '=' when expecting identifier following 'int.'
../.dub/packages/enumap-0.4.0/src/enumap.d:222: error: semicolon expected following auto declaration, not '0'
../.dub/packages/enumap-0.4.0/src/enumap.d:249: error: user defined attributes cannot appear as postfixes
../.dub/packages/enumap-0.4.0/src/enumap.d:249: error: unexpected identifer 'dg' in declarator
../.dub/packages/enumap-0.4.0/src/enumap.d:249: error: found ';' when expecting ')'
../.dub/packages/enumap-0.4.0/src/enumap.d:251: error: found 'res' when expecting '.' following int
../.dub/packages/enumap-0.4.0/src/enumap.d:251: error: found '=' when expecting identifier following 'int.'
../.dub/packages/enumap-0.4.0/src/enumap.d:251: error: semicolon expected following auto declaration, not '0'
../.dub/packages/enumap-0.4.0/src/enumap.d:452: error: user defined attributes cannot appear as postfixes
../.dub/packages/enumap-0.4.0/src/enumap.d:452: error: semicolon expected following function declaration
../.dub/packages/enumap-0.4.0/src/enumap.d:452: error: Declaration expected, not '('
../.dub/packages/enumap-0.4.0/src/enumap.d:473: error: function declaration without return type. (Note that constructors are always named 'this')
../.dub/packages/enumap-0.4.0/src/enumap.d:473: error: no identifier for declarator helper(pairs)
../.dub/packages/enumap-0.4.0/src/enumap.d:474: error: Declaration expected, not 'return'
../.dub/packages/enumap-0.4.0/src/enumap.d:475: error: unrecognized declaration
gdc failed with exit code 1.
```

**Comment by rcorre** [comment]
That's a weird restriction, but should be easy enough to fix. 
I have a raspberry pi I could test on, though I've never used it to compile D.

Thanks for reporting!

**Comment by rcorre** [comment]
I tried compiling with gdc `5.3.0` and didn't get complaints about the order of
attribute declarations, so maybe its only a problem with older GDC versions?
This wasn't on my ARM device but it seems unlikely that would be the issue.

**Comment by SimonN** [comment]
Thanks for investigating! It's likely that gdc has improved on this between 4.9 and 5.3. I too don't think it's got anything to do with ARM.

Since enumap builds nicely with a modern gdc, I believe you need not dig deeper. Instead, I should try to fix the build errors when I get the chance and make a PR. Then you can decide what to do.

**Comment by rcorre** [comment]
>  Since enumap builds nicely with a modern gdc

It does build with a modern GDC, but the tests tell another story.
Apparently GDC and DMD [disagree on where @nogc can be applied](http://forum.dlang.org/thread/pajldkpyopfosqkptxps@forum.dlang.org#post-pajldkpyopfosqkptxps:40forum.dlang.org). Its a separate issue, but I'm glad you got me to try GDC :)

**Comment by rcorre** [comment]
@SimonN I pushed `v0.4.1` which addresses the attribute ordering as well as other issues trying GDC brought up.

**Comment by SimonN** [comment]
0.4.1 builds perfectly with DMD, both natively on Linux and via Wine. I haven't gotten around to test it with GDC on ARM, but will inform you once I do. Thanks so far!

**Comment by SimonN** [comment]
Here are the results of enumap 0.4.1 from GDC 4.9.2 on ARM.

```
enumap.d:222: error: basic type expected, not @
enumap.d:222: error: semicolon expected to close alias declaration
enumap.d:222: error: no identifier for declarator int delegate(K, const V)
enumap.d:222: error: use '{ }' for an empty statement, not a ';'
enumap.d:252: error: basic type expected, not @
enumap.d:252: error: semicolon expected to close alias declaration
enumap.d:252: error: no identifier for declarator int delegate(K, ref V)
enumap.d:252: error: use '{ }' for an empty statement, not a ';'
```

So, GDC doesn't like the @nogc in front of the function, whereas last week, it didn't like @nogc after the function.

Line 222 aliases a function type, instead of declaring a function. Maybe that was playing a role. We tried `dub -b unittest`, and the first error was again the above error in line 222, even though there are `@nogc unittest {` earlier in the file.

As a workaround, we replaced all `@nogc` with empty string. That got rid of all errors, and enumap worked. As I write this, I get another idea; I should have tested `@nogc` between type and function name.

Because you have tried enumap 0.4.1 with a modern GDC version, and it builds for you, I'd blame the remaining problem on GDC 4.9.2's age.

According to the [GDC wiki](http://wiki.dlang.org/GDC), the cutting-edge GDC 5.2 uses the DMD 2.066 frontend, which is the same frontend used by the above-tested GDC 4.9.2. That makes me wonder why the build doesn't pass...

**Comment by rcorre** [comment]
> Line 222 aliases a function type, instead of declaring a function

The original line was

```
auto callme = cast(int delegate(K, ref V) @nogc) dg;
```

Since it seemed like GDC didn't like the `@nogc` postfix, I moved it to a
prefix. 

```
auto callme = cast(@nogc int delegate(K, ref V)) dg;
```

However, DMD didn't seem to like having a prefix attribute in a cast, but didn't
complain if I did:

```
alias nogcDelegate = @nogc int delegate(K, ref V);
auto callme = cast(nogcDelegate) dg;
```

Which seemed to work fine on modern DMD and GDC.
If GDC is still complaining, I'm not quite sure what it wants here.
That being said, this whole 'cast to @nogc' busisness is rather messy, but
necessary if the caller wants to `foreach` in an `@nogc` context (as `@nogc`
isn't inferred for the delegate).
